duckduckgo/macos-browser
# Fix Website Breakage Reports by allowing unescaped commas in URL query

Task/Issue URL: https://app.asana.com/0/1163321984198618/1202033667108435/f
Tech Design URL:
CC:

**Description**:
Website Breakage reporting uses comma (a _reserved character_ as described by [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#section-2.2)) to separate blocked trackers domains in a single URL query parameter. Since we upgraded URL.addParameter to percent-encode all reserved characters, macOS website breakage reports got malformed, because of percent-encoded commas.
This change utilizes an update to `URL.addParameter` (added in BSK) that allows to explicitly specify characters that are excluded from percent encoding. Then, in `WebsiteBreakageSender`, a comma is added as allowed character, to fix the reports.
Adding this new API required updating `APIRequest` and `Pixel`'s code, but since the allowed characters set defaults to `nil` and is an opt-in feature, no API request other than Website Breakage should change.

I also added the initial version of Website Breakage reference tests, which checks contents of URL requests sent via Pixel. [Here is the reference JSON file I was inspired by](https://github.com/duckduckgo/privacy-reference-tests/blob/main/broken-site-reporting/tests.json). Since it turned out that our breakage reporting mechanism is a little bit off relative to the reference, I made the following changes:
* removed query and fragment from website URLs
* added the `gpc` flag
* parametrized the `manufacturer` - it defaults to "Apple" anyway, but reference tests JSON contains different values there so I enabled it for the future when we'll have automatic tests working on the JSON file.

**Steps to test this PR**:
1. Make sure all tests pass
1. Make sure semicolons in search queries work as intended (i.e. are percent-encoded)
1. Make sure feedback is not truncated at the semicolon
1. Verify that all other network connectivity (that I'm not aware of yet) besides web browsing is intact

**Testing checklist**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**
